torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  TORCH_INTERNAL_ASSERT(
+      out_tv->memoryType() == MemoryType::Global ||
+          in_tv->memoryType() == MemoryType::Global,
+      "Either input or output tensor must be on global memory.");
+  // However, not both of them
+  TORCH_INTERNAL_ASSERT(
+      !(out_tv->memoryType() == MemoryType::Global &&
+        in_tv->memoryType() == MemoryType::Global),
+      "Both input and output tensors are on global memory.");
+  // Must be either global or local
+  TORCH_INTERNAL_ASSERT(
+      (out_tv->memoryType() == MemoryType::Global ||
+       out_tv->memoryType() == MemoryType::Local),
+      "Invalid memory type of output tensor");
+  TORCH_INTERNAL_ASSERT(
+      (in_tv->memoryType() == MemoryType::Global ||
+       in_tv->memoryType() == MemoryType::Local),
+      "Invalid memory type of input tensor");
+
+  // TensorView on global memory. This is the tensor that may have
+  // a non-aligned base address.
+  auto global_tv =
+      (out_tv->memoryType() == MemoryType::Global) ? out_tv : in_tv;
+
+  // TensorView with the misaligned vec iterDomain. It is the consumer
+  // of vectorized load or the producer of vectorized store. It is
+  // assumed that when the output TV is not on global memory, this
+  // expression is a vectorized load, so the output TV is vec_tv.
+  // TODO: Check vec_tv has indeed MisalignedVectorize parallel type.
+  auto vec_tv = (out_tv->memoryType() != MemoryType::Global) ? out_tv : in_tv;
+
+  auto pred = PredicateCompute::getInlinePredicate(
+      vec_expr, loop_structure, nullptr, false, true);
+  if (pred == nullptr) {
+    pred = ir_builder.create<kir::Bool>(true);
+  }
+
+  kir::IfThenElse* pred_ite = ir_builder.create<kir::IfThenElse>(pred);
+  new_loop->body().push_back(pred_ite);
+
+  // Generate vectorize index
+  auto indices = (out_tv->memoryType() == MemoryType::Global)
+      ? Index::getConsumerStridedIndices(out_tv->fuserTv(), loop_structure)
+      : Index::getProducerStridedIndices(
+            in_tv->fuserTv(), out_tv->fuserTv(), loop_structure);
+
+  //  Get full extent for merged root domains
+  auto extent = getVectorizeExtent(vec_tv, indices);
+
+  auto vector_size =
+      vec_tv->domain()->domain().back()->extent()->as<kir::Int>();
+
+  auto index =
+      ir_builder.create<kir::TensorIndex>(global_tv->fuserTv(), indices);
+  auto base_address =
+      setupNamedScalar(pred_ite->thenBody(), index, "base_address", true);
+
+  kir::Int* data_size =
+      ir_builder.create<kir::Int>(dataTypeSize(vec_tv->dtype()));
+  auto vector_data_size = ir_builder.mulExpr(vector_size, data_size);
+  auto a = ir_builder.modExpr(base_address, vector_data_size);
+  auto b = ir_builder.divExpr(a, data_size);
+  auto c = ir_builder.subExpr(vector_size, b);
+  auto shift_init = setupNamedScalar(pred_ite->thenBody(), c, "shift_val");
+
+  auto shift_pred = ir_builder.eqExpr(shift_init, vector_size);
+  auto shift_val =
+      ir_builder.whereExpr(shift_pred, ir_builder.zero(), shift_init);
```

Comment:
I'd really like to understand why we're using a `whereExpr` here, this seems odd to me.

Reply:
`shift = vec_size - (address % vec_size)`
The range of shift is [1, vec_size]. However, this can cause an issue with broadcast tensors.
The where clause maps the range to [0, vec_size-1].
`shift = (shift == vec_size) ? 0 : shift`